File: days/src/main/scala/aoc2024s/day9/Day9.scala

package aoc2024s.day9;

import utils.IO

import scala.annotation.tailrec
import scala.jdk.CollectionConverters.*

object Day9 {

  def checksum(id: Int, start: Long, size: Long): Long =
    id * ((2 * start + size - 1) * size) / 2;

  case class FileBlock(id: Int, start: Long, size: Long) {
    def checksum: Long = Day9.checksum(id, start, size);
  }

  case class FreeBlock(start: Long, size: Long)

  def fromDense(dense: String): (List[FileBlock], List[FreeBlock]) = {
    val lengths = dense.split("").map(Integer.parseInt).toList
    val starts = lengths.scanLeft(0L)(_ + _)
    val initial = starts.zip(lengths).zipWithIndex
    val (files, free) = initial.partition(_._2 % 2 == 0)
    val fileBlocks = files.map { case ((start, size), id) => FileBlock(id / 2, start, size) }
    val freeBlocks = free.map { case ((start, size), _) => FreeBlock(start, size) }
    (fileBlocks, freeBlocks)
  }

  @tailrec
  final def part1(fileBlocks: List[FileBlock], freeBlocks: List[FreeBlock], checksum: Long = 0L): Long = {
    fileBlocks match {
      case Nil => checksum
      case (lastFile@FileBlock(id, fileStart, fileSize)) :: restFiles =>
        freeBlocks match
          case Nil => checksum + fileBlocks.map(_.checksum).sum
          case FreeBlock(freeStart, freeSize) :: restFree =>
            if freeStart >= fileStart then
              checksum + fileBlocks.map(_.checksum).sum
            else if fileSize <= freeSize then
              val newChecksum = checksum + Day9.checksum(id, freeStart, fileSize)
              val newFreeBlock = FreeBlock(freeStart + fileSize, freeSize - fileSize)
              if fileSize == freeSize then
                part1(restFiles, restFree, newChecksum)
              else
                part1(restFiles, newFreeBlock :: restFree, newChecksum)
            else
              val restSize = fileSize - freeSize
              val newChecksum = checksum + Day9.checksum(id, freeStart, freeSize)
              val newFileBlock = FileBlock(id, fileStart, restSize)
              part1(newFileBlock :: restFiles, restFree, newChecksum)
    }
  }

  @tailrec
  final def part2(fileBlocks: List[FileBlock], freeBlocks: Vector[FreeBlock], checksum: Long = 0L): Long = {
    fileBlocks match {
      case Nil => checksum
      case (lastFile@FileBlock(id, fileStart, fileSize)) :: restFiles =>
        freeBlocks.indexWhere(fileSize <= _.size) match
          case -1 => part2(restFiles, freeBlocks, checksum + lastFile.checksum)
          case i =>
            val FreeBlock(freeStart, freeSize) = freeBlocks(i)
            if freeStart >= fileStart then
              part2(restFiles, freeBlocks, checksum + lastFile.checksum)
            else
              val newChecksum = checksum + Day9.checksum(id, freeStart, fileSize)
              val newFreeBlock = FreeBlock(freeStart + fileSize, freeSize - fileSize)
              if fileSize == freeSize then
                part2(restFiles, freeBlocks.patch(i, Nil, 1), newChecksum)
              else
                part2(restFiles, freeBlocks.updated(i, newFreeBlock), newChecksum)
    }
  }

  def part1(data: List[String]): Long = {
    val (fileBlocks, freeBlocks) = fromDense(data.head)
    part1(fileBlocks.reverse, freeBlocks)
  }

  def part2(data: List[String]): Long = {
    val (fileBlocks, freeBlocks) = fromDense(data.head)
    part2(fileBlocks.reverse, freeBlocks.toVector)
  }

  @main def main9(): Unit = {
    val data = IO.getResourceAsList("aoc2024/day9.txt").asScala.toList;
    val part1 = Day9.part1(data);
    println(s"part1 = $part1");
    val part2 = Day9.part2(data);
    println(s"part2 = $part2");
  }
}

File: days/src/main/java/aoc2024j/day9/Day9.java

package aoc2024j.day9;

import utils.IO;

import java.util.*;
import java.util.function.IntConsumer;

public class Day9 {

    record Block(long start, long size) {
        long checksum() {
            return ((2 * start + size - 1) * size) / 2;
        }
    }

    static class FileSystem {
        final LinkedHashMap<Integer, List<Block>> files = new LinkedHashMap<>();
        final List<Block> free = new ArrayList<>(); // free blocks with leftmost as last

        static FileSystem fromDense(String dense) {
            var fs = new FileSystem();
            int start = 0;
            for (int i = 0; i < dense.length(); i++) {
                int size = Integer.parseInt(dense.substring(i, i + 1));
                var block = new Block(start, size);
                if (size == 0) continue;
                if (i % 2 == 0) {
                    var blocks = new ArrayList<Block>();
                    blocks.add(block);
                    fs.files.put(i / 2, blocks);
                } else {
                    fs.free.add(block);
                }
                start += size;
            }
            Collections.reverse(fs.free); // leftmost block is last
            return fs;
        }

        void compactFileSystem(IntConsumer fileCompactor) {
            for (var entry : files.sequencedEntrySet().reversed()) {
                var id = entry.getKey();
                if (!free.isEmpty()) {
                    fileCompactor.accept(id);
                }
            }
        }

        private void compactBlocks(Integer id) {
            long fileStart = files.get(id).getFirst().start;
            long fileSize = files.get(id).getFirst().size;
            var newFileBlocks = new ArrayList<Block>();
            while (fileSize > 0 && !free.isEmpty() && free.getLast().start < fileStart) {
                var freeBlock = free.removeLast();
                var newFreeBlock = new Block(freeBlock.start + fileSize, freeBlock.size - fileSize);
                if (newFreeBlock.size > 0) free.addLast(newFreeBlock);
                var newFileBlock = new Block(freeBlock.start, Math.min(fileSize, freeBlock.size));
                newFileBlocks.add(newFileBlock);
                fileSize -= newFileBlock.size;
            }
            if (fileSize > 0) {
                var newFileBlock = new Block(fileStart, fileSize);
                newFileBlocks.add(newFileBlock);
            }
            files.put(id, newFileBlocks);
        }

        private void compactWholeFiles(Integer id) {
            long fileStart = files.get(id).getFirst().start;
            long fileSize = files.get(id).getFirst().size;
            for (int i = free.size() - 1; i >= 0; i--) {
                var freeBlock = free.get(i);
                if (freeBlock.start > fileStart) continue;
                if (freeBlock.size >= fileSize) {
                    var newFreeBlock = new Block(freeBlock.start + fileSize, freeBlock.size - fileSize);
                    if (newFreeBlock.size > 0) free.set(i, newFreeBlock);
                    else free.remove(i);
                    var newFileBlock = new Block(freeBlock.start, fileSize);
                    files.put(id, List.of(newFileBlock));
                    break;
                }
            }
        }

        long checksum() {
            return files.entrySet().stream()
                    .mapToLong(entry ->
                            entry.getValue().stream()
                                    .mapToLong(Block::checksum).sum() * entry.getKey())
                    .sum();
        }
    }

    long part1(List<String> data) {
        var fs = FileSystem.fromDense(data.getFirst());
        fs.compactFileSystem(fs::compactBlocks);
        return fs.checksum();
    }

    long part2(List<String> data) {
        var fs = FileSystem.fromDense(data.getFirst());
        fs.compactFileSystem(fs::compactWholeFiles);
        return fs.checksum();
    }

    public static void main() {
        var day9 = new Day9();
        var data = IO.getResourceAsList("aoc2024/day9.txt");
        var part1 = day9.part1(data);
        System.out.println("part1 = " + part1);
        var part2 = day9.part2(data);
        System.out.println("part2 = " + part2);
    }
}
